src/amber/controller/helpers/responders.cr

```diff
@@ -61,9 +66,12 @@ module Amber::Controller::Helpers
     end
 
     def set_response(body, status_code = 200, content_type = Content::TYPE[:html])
-      context.response.status_code = status_code
+      if body.is_a?(Proc)
```

Comment:
body should be resolved within the instance of content instead of inside of set response. `set_response` should have the Single Responsibility of setting the values passed into it.

```cr
#body inside of Content
       def body
         _body = @available_responses[type]?
         if _body.is_a?(Proc)
           _body = _body.call
         end
        return _body
       end
```

Reply:
The reason the Proc.call is in set_response is that if it is Content.body then the Proc will be called twice in respond_with.  Once to test if it exists and again to pass it to set_response.  Calling the Proc in set_response delays the execution of the Proc until the last method call in the request handle process.

Reply:
I believe a case statement will clean this up a bit

```crystal 
case _body = @available_responses[type]?
when Proc then _body.call
else _body
end
```

What do you guys think?

Reply:
@damianham Sorry I'm responding to this so late.

Body could easily be cached as `@body` which would stop it from being resolved multiple times.

Reply:
yes good point